File: ch02/nodejs/2.4.1.findSubstring.js
const findSubstringIter = (needle, haystack) => {
    for (let i = 0; i < haystack.length; i++) {
        if (haystack.slice(i, i + needle.length) === needle) return i
    }

    return -1
}

console.log(findSubstringIter('cat', 'My cat zophie'))

const findSubstringRecur = (needle, haystack, i = 0) => {
    if (i > haystack.length) return -1;

    if (haystack.slice(i, needle.length + i) === needle) return i;

    return findSubstringRecur(needle, haystack, i + 1);
}

console.log(findSubstringRecur('cat', 'My cat zophie'))

File: ch02/python/2.4.1.find_substring.py
def find_substring_iter(needle, haystack):
    i = 0

    while (i < len(haystack)):
        substr = haystack[i : i + len(needle)]
        print(substr)
        if substr == needle:
            return i
        i += 1

    return -1

i = find_substring_iter('cat', 'My cat zophie')
print(i)

def find_substring_recur(needle, haystack, i = 0):

    if i > len(haystack):
        return -1

    if haystack[i : i + len(needle)] == needle:
        return i

    return find_substring_recur(needle, haystack, i + 1)


i = find_substring_recur('cat', 'My cat zophie')
print(i)